src/steps/services.ts

```diff
@@ -0,0 +1,184 @@
+    const serviceEntity = createIntegrationEntity({
+      entityData: {
+        source: data,
+        assign: {
+          _type: 'fastly_service',
+          _class: 'Service',
+          _key: `fastly-service:${data.id}`,
+          id: data.id,
+          displayName: data.name || data.id,
+          name: data.name,
+          category: ['infrastructure'],
+          description: data.comment,
+          version: data.version,
+          createdOn: getTime(data.created_at),
+          updatedOn: getTime(data.updated_at),
+        },
+      },
+    });
+
+    await Promise.all([
+      jobState.addEntity(serviceEntity),
+      jobState.addRelationship(
+        createDirectRelationship({
+          _class: RelationshipClass.HAS,
+          fromKey: `fastly-account:${instance.config.customerId}`,
+          fromType: 'fastly_account',
+          toKey: `fastly-service:${data.id}`,
+          toType: 'fastly_service',
+        }),
+      ),
+    ]);
+
+    const backends = await apiClient.getServiceBackends(data.id, data.version);
+
+    for (const backend of backends) {
+      const backendEntity = createIntegrationEntity({
+        entityData: {
+          source: backend,
+          assign: {
+            ...convertProperties(backend),
+            _type: 'fastly_service_backend',
+            _class: 'ApplicationEndpoint',
+            _key: `fastly-service-backend:${backend.hostname}:${backend.address}:${backend.port}`,
+            displayName: backend.name || backend.hostname || backend.address,
+            description: backend.comment,
+            ipAddress: backend.address,
+            createdOn: getTime(backend.created_at),
+            updatedOn: getTime(backend.updated_at),
+            deletedOn: getTime(backend.deleted_at),
+          },
+        },
+      });
+
+      await Promise.all([
+        jobState.addEntity(backendEntity),
+        jobState.addRelationship(
+          createDirectRelationship({
+            _class: RelationshipClass.HAS,
+            from: serviceEntity,
+            to: backendEntity,
+          }),
+        ),
+        jobState.addRelationship(
+          createMappedRelationship({
+            _type: 'fastly_service_backend_connects_host',
+            _class: RelationshipClass.CONNECTS,
+            _mapping: {
+              sourceEntityKey: backendEntity._key,
+              relationshipDirection: RelationshipDirection.FORWARD,
+              targetFilterKeys: [
+                ['_class', 'ipAddress'],
+                ['_class', 'hostname'],
+              ],
+              targetEntity: {
+                _class: ['Host', 'Gateway'],
+                ipAddress: backend.address,
+                hostname: backend.hostname,
+              },
+              skipTargetCreation: true,
+            },
+          }),
+        ),
+      ]);
+    }
+
+    const domains = await apiClient.getServiceDomains(data.id, data.version);
+
+    for (const domain of domains) {
+      await Promise.all([
```

Comment:
This is only a `Promise.all` call with a single promise, so it is semantically equivalent to just awaiting on the single promise.

Additionally, I think we should generally try to avoid adding data to the job state in parallel. I know it isn't clear, but under the hood the job state collects entities/relationships and uploads them in batches already.

Reply:
will update.

if we want to discourage `Promise.all`, we should update the template accordingly: https://github.com/JupiterOne/integration-template/blob/master/src/steps/access.ts#L39

Reply:
Created https://github.com/JupiterOne/sdk/issues/323 to track a couple things we need to do to the template project. Thanks @erkangz !